### PR TITLE
add command shell blacklist and whitelist

### PR DESCRIPTION
**Pull request by Wladastic**
### Background

There is only a command category whitelist and blacklist.

### Changes

Added the ability to restrict shell_commands by keywords, to avoid that the instance keeps running into ui interfaces that have no return.

### Documentation

Add your list of keywords to be forbidden.
When you add a keyword to the whilelist, only those keywords from the whitelist are allowed.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [ ] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -35,6 +35,15 @@
 ## For example, to disable coding related features, uncomment the next line
 # DISABLED_COMMAND_CATEGORIES=autogpt.commands.analyze_code,autogpt.commands.execute_code,autogpt.commands.git_operations,autogpt.commands.improve_code,autogpt.commands.write_tests
 
+## BLACKLISTED_COMMANDS - The list of commands that are not allowed to be executed by Auto-GPT (Default: None)
+# the following are examples:
+# BLACKLISTED_COMMANDS=cd,nano,vim,vi,emacs,rm,sudo,top,ping,ssh,scp
```
Comment (richbeales) [inline .env.template]: Please switch terminology to allow/deny

Comment (Wladastic) [reply .env.template]: Hm, sounds way better

File: autogpt/commands/execute_code.py [hunk 2]
```diff
@@ -97,6 +97,30 @@ def execute_python_file(filename: str) -> str:
...
+
+    if not tokens:
+        return False
+
+    if CFG.deny_commands and tokens[0] not in CFG.deny_commands:
+        return False
+
+    for keyword in CFG.allow_commands:
+        if keyword in tokens:
+            return False
```
Comment (k-boikov) [inline autogpt/commands/execute_code.py]: Shouldnt that be return True?

**Comment by lc0rp** [comment]
Echoing comment from: https://github.com/Significant-Gravitas/Auto-GPT/pull/2644

To promote more inclusive language within the codebase, this pull request addresses the usage of terms such as "blacklist" and "whitelist." The new terms "denylist" and "allowlist" were been proposed as more neutral alternatives.

**Comment by Wladastic** [comment]
> Echoing the sentiment from #2644.
> 
> The proposal to use "denylist" and "allowlist" instead of "blacklist" and "whitelist" was made to encourage more inclusive language in the codebase. These terms are considered more neutral alternatives.

In the code it says allow and deny already.
Are you implying to use denylist and allowlist as well here?
It makes less sense in that context I think

File: autogpt/commands/execute_code.py [hunk 3]
```diff
@@ -97,6 +97,30 @@ def execute_python_file(filename: str) -> str:
...
+
+    if not tokens:
+        return False
+
+    if CFG.deny_commands and tokens[0] not in CFG.deny_commands:
+        return False
+
+    for keyword in CFG.allow_commands:
+        if keyword in tokens:
+            return True
```
Comment (k-boikov) [inline autogpt/commands/execute_code.py:119]: Now looks better, but it is still missing logic - if you have allowlist (not empty) and it doesnt find the command in the allowlist it must return False.

**Review by k-boikov — CHANGES_REQUESTED** [review]

File: autogpt/commands/execute_code.py (hunk 3 above)
Comment (Wladastic) [reply autogpt/commands/execute_code.py:119]: I feel like there should be a prettier way than I did...

**Review by k-boikov — APPROVED** [review]

**Review by richbeales — APPROVED** [review]

**Pull request merged**